search2/python/search/utils/beanstalkd.py

```diff
@@ -0,0 +1,199 @@
+from glob import glob
+from os import path
+import time
+import traceback
+from typing import Any, NamedTuple, List, Callable, Optional
+
+import asyncio
+from orjson import loads, dumps  # pylint: disable=no-name-in-module
+from pystalk import BeanstalkClient, BeanstalkError
+from opensearchpy.exceptions import NotFoundError
+
+import ray
+
+BEANSTALK_ERR_TIMEOUT = "TIMED_OUT"
+
+class Message(NamedTuple):
+    """Beanstalkd Message"""
+    event: str
+    queueID: str
+    submissionID: str
+    data: Any = None
+
+class Publisher(NamedTuple):
+    """Beanstalkd Message Published Config"""
+    host: str
+    port: str
+    queue: str
+    message: Message
+
+class QueueConf(NamedTuple):
+    """Queue Configuration"""
+    addresses: List[str]
+    events: dict
+    tubes: dict
+
+    def split_host_port(self):
+        """Split host and port"""
+        hosts = []
+        ports = []
+        for host in self.addresses:
+            host, port = host.split(":")
+            hosts.append(host)
+            ports.append(port)
+        return hosts, ports
+
+def get_config_file_path(config_path_base_dir: str, assembly: str, suffix: str = ".y*ml"):
+    """Get config file path"""
+    paths = glob(path.join(config_path_base_dir, assembly + suffix))
+
+    if not paths:
+        raise ValueError(f"\n\nNo config path found for the assembly {assembly}. Exiting\n\n")
+
+    if len(paths) > 1:
+        print("\n\nMore than 1 config path found, choosing first")
+
+    return paths[0]
+
+def _specify_publisher(
+    submission_id, job_id, publisher_host, publisher_port, progress_event, event_queue
+):
+    return Publisher(
+            host = publisher_host,
+            port = publisher_port,
+            queue = event_queue,
+            message = Message(
+                event = progress_event,
+                queueID = job_id,
+                submissionID = submission_id,
+                data = None,
+            ))
+
+def listen(
+    handler_fn: Callable,
+    submit_msg_fn: Callable,
+    completed_msg_fn: Callable,
+    queue_conf: QueueConf,
+    tube: str,
+):
+    """Listen on a Beanstalkd channel, waiting for work.
+       When work is available call the work handler
+    """
+    hosts, ports = queue_conf.split_host_port()
+
+    for event in ("progress", "failed", "started", "completed"):
+        assert event in queue_conf.events
+
+    tube_conf = queue_conf.tubes[tube]
+    events_conf = queue_conf.events
+    clients = tuple(BeanstalkClient(host, port, socket_timeout=10) for (host, port) in zip(hosts, ports))
+
+    i = 0
+    while True:
+        i += 1
+        offset = i % len(hosts)
+        client  = clients[offset]
+
+        client.watch(tube_conf["submission"])
+
+        try:
+            job = client.reserve_job(5)
+        except BeanstalkError as err:
+            if err.message == BEANSTALK_ERR_TIMEOUT:
+                continue
+            raise err
+        try:
+            job_data = loads(job.job_data)
```

Comment:
this should really be its own try-block.  we're deserializing json that came in over telnet and that could really be anything.  We can validate immediately here and bail if it's malformed, raising an informative exception.